File: sdk/php/src/Providers/Anonymmail.php
<?php

declare(strict_types=1);

namespace ChanhanzhanX\TempMail\Providers;

use ChanhanzhanX\TempMail\Email;
use ChanhanzhanX\TempMail\EmailInfo;
use ChanhanzhanX\TempMail\HttpClient;
use ChanhanzhanX\TempMail\Normalize;

/**
 * Anonymmail 渠道实现 — https://anonymmail.net
 *
 * POST 表单 API：先取域名，本地拼随机用户名，POST /api/create 创建；
 * 读信 POST /api/get，响应形如 {"email@domain": {"emails": [...]}}。
 */
final class Anonymmail
{
    private const CHANNEL = 'anonymmail';
    private const BASE = 'https://anonymmail.net';

    /** @var array<string,string> */
    private const HEADERS = [
        'Accept' => '*/*',
        'Content-Type' => 'application/x-www-form-urlencoded',
        'Referer' => 'https://anonymmail.net/',
        'User-Agent' => 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 '
            . '(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0',
    ];

    private static function randomUsername(int $length = 10): string
    {
        $chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
        $out = '';
        for ($i = 0; $i < $length; $i++) {
            $out .= $chars[random_int(0, strlen($chars) - 1)];
        }
        return $out;
    }

    /**
     * 获取可用域名列表。
     *
     * @return string[]
     */
    private static function fetchDomains(): array
    {
        $resp = HttpClient::post(self::BASE . '/api/getDomains', self::HEADERS);
        $data = HttpClient::json($resp);
        if (!array_is_list($data)) {
            return [];
        }
        $out = [];
        foreach ($data as $item) {
            if (is_array($item)) {
                $domain = trim((string) ($item['domain'] ?? ''));
                if ($domain !== '') {
                    $out[] = $domain;
                }
            }
        }
        return $out;
    }

    public static function generate(): EmailInfo
    {
        $domains = self::fetchDomains();
        if ($domains === []) {
            throw new \RuntimeException('anonymmail: no domains available');
        }
        $domain = $domains[array_rand($domains)];
        $email = self::randomUsername() . '@' . $domain;

        // HEAD 请求获取 cookie session（本端共享客户端自动持有 Cookie）
        HttpClient::get(self::BASE . '/', self::HEADERS);

        $resp = HttpClient::post(self::BASE . '/api/create', self::HEADERS, body: 'email=' . $email);
        if ($resp->getStatusCode() >= 400) {
            throw new \RuntimeException('anonymmail: create inbox failed ' . $resp->getStatusCode());
        }
        $data = HttpClient::json($resp);
        if (empty($data['success'])) {
            throw new \RuntimeException('anonymmail: create inbox failed');
        }
        $addr = trim((string) ($data['email'] ?? ''));
        if ($addr === '') {
            throw new \RuntimeException('anonymmail: missing email in response');
        }
        return new EmailInfo(self::CHANNEL, $addr);
    }

    /**
     * @return Email[]
     */
    public static function getEmails(string $email): array
    {
        $e = trim($email);
        if ($e === '') {
            throw new \InvalidArgumentException('anonymmail: empty email');
        }
        $resp = HttpClient::post(self::BASE . '/api/get', self::HEADERS, body: 'email=' . $e);
        if ($resp->getStatusCode() >= 400) {
            throw new \RuntimeException('anonymmail: 收件箱请求失败 ' . $resp->getStatusCode());
        }
        $data = HttpClient::json($resp);
        // 响应格式: {"email@domain": {"created_at": "...", "emails": [...]}}
        $inbox = $data[$e] ?? null;
        if (!is_array($inbox)) {
            return [];
        }
        $rows = $inbox['emails'] ?? null;
        if (!is_array($rows)) {
            return [];
        }
        $out = [];
        foreach ($rows as $raw) {
            if (!is_array($raw)) {
                continue;
            }
            $normalized = $raw;
            // token 字段映射为 id
            if (isset($normalized['token']) && !isset($normalized['id'])) {
                $normalized['id'] = $normalized['token'];
                unset($normalized['token']);
            }
            // body 字段映射为 html（body 含 HTML 内容）
            if (isset($normalized['body']) && !isset($normalized['html'])) {
                $normalized['html'] = $normalized['body'];
                unset($normalized['body']);
            }
            $out[] = Normalize::email($normalized, $e);
        }
        return $out;
    }
}

File: sdk/csharp/src/Providers/Anonymmail.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Text.Json.Nodes;

namespace XxxXTeam.TempMail.Providers;

/// <summary>
/// Anonymmail 渠道（anonymmail.net）。
/// POST 表单 API，需先建立会话 Cookie（共享 HttpClient 内置 CookieContainer 自动维护）。
/// </summary>
public static class Anonymmail
{
    private const string Base = "https://anonymmail.net";
    private static readonly Random Rand = new();

    private static readonly Dictionary<string, string> Headers = new()
    {
        ["Accept"] = "*/*",
        ["Content-Type"] = "application/x-www-form-urlencoded",
        ["Referer"] = "https://anonymmail.net/",
        ["User-Agent"] = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0",
    };

    private static string RandomUsername(int length = 10)
    {
        const string chars = "abcdefghijklmnopqrstuvwxyz0123456789";
        var sb = new System.Text.StringBuilder();
        for (var i = 0; i < length; i++) sb.Append(chars[Rand.Next(chars.Length)]);
        return sb.ToString();
    }

    private static List<string> FetchDomains()
    {
        var resp = Http.Post($"{Base}/api/getDomains", "", "application/x-www-form-urlencoded", Headers);
        resp.EnsureSuccess();
        var arr = Json.Parse(resp.Body) as JsonArray;
        var domains = new List<string>();
        if (arr is null) return domains;
        foreach (var item in arr)
        {
            if (item is not JsonObject o) continue;
            var d = (o["domain"]?.GetValue<string>() ?? "").Trim();
            if (d.Length > 0) domains.Add(d);
        }
        return domains;
    }

    /// <summary>创建临时邮箱：先 GET 首页获取会话 Cookie，再 POST /api/create</summary>
    public static EmailInfo Generate()
    {
        var domains = FetchDomains();
        if (domains.Count == 0) throw new Exception("anonymmail: no domains available");
        var domain = domains[Rand.Next(domains.Count)];
        var email = $"{RandomUsername()}@{domain}";

        // GET 首页建立会话 Cookie（共享客户端的 CookieContainer 自动携带）
        try { Http.Get($"{Base}/", Headers); } catch { /* 忽略预热失败 */ }

        var resp = Http.Post($"{Base}/api/create", $"email={email}", "application/x-www-form-urlencoded", Headers);
        resp.EnsureSuccess();
        var data = Json.Parse(resp.Body) as JsonObject;
        if (data is null || (data["success"] as JsonValue)?.GetValue<bool>() != true)
            throw new Exception("anonymmail: create inbox failed");
        var addr = (data["email"]?.GetValue<string>() ?? "").Trim();
        if (addr.Length == 0) throw new Exception("anonymmail: missing email in response");
        return new EmailInfo("anonymmail", addr);
    }

    /// <summary>获取邮件列表</summary>
    public static List<Email> GetEmails(string email)
    {
        var e = (email ?? "").Trim();
        if (e.Length == 0) throw new Exception("anonymmail: empty email");

        var resp = Http.Post($"{Base}/api/get", $"email={e}", "application/x-www-form-urlencoded", Headers);
        resp.EnsureSuccess();
        var data = Json.Parse(resp.Body) as JsonObject;
        var result = new List<Email>();
        if (data is null) return result;
        // 响应格式: {"email@domain": {"created_at": "...", "emails": [...]}}
        if (data[e] is not JsonObject inbox) return result;
        if (inbox["emails"] is not JsonArray rows) return result;

        foreach (var raw in rows)
        {
            if (raw is not JsonObject ro) continue;
            var dict = Json.ToDict(ro);
            // token 映射为 id
            if (dict.TryGetValue("token", out var tok) && !dict.ContainsKey("id"))
            {
                dict["id"] = tok;
                dict.Remove("token");
            }
            // body 映射为 html（body 含 HTML 内容）
            if (dict.TryGetValue("body", out var body) && !dict.ContainsKey("html"))
            {
                dict["html"] = body;
                dict.Remove("body");
            }
            result.Add(Normalize.NormalizeEmail(dict, e));
        }
        return result;
    }
}
